### PR TITLE
Nuget package

### PR DESCRIPTION
**Pull request by vpenades**
As discussed in #134, In this pull request I dropped the basic idea of a nuget package.

Its basic usage would be this:

```c#
var report = glTF2Validator.ValidationContext.ValidateFile("somefile.gltf");

if (report.HasErrors) throw new Exception(report.RawReport);
```

In order to compile the project, it is required to compile the CLI tools beforehand, since they're embedded in the nuget package.

I have only tested the Windows path; theoretically it should work on Linux, but someone will need to do the testing.

I haven't defined anything for the semantic versioning. Ideally, the nuget package should have the same version ID as the gltf-validator itself.

### CODE REVIEW

**Comment by vpenades** [comment]
@lexaknyazev so far, what's left is:

- Amend the csproj with copyright/license/author stuff, and I guess setting the output path for the package.
- Setup an automated build mechanism that builds this project after the CLI tools have been built, that also passes the semantic version.

I believe it's better to leave these last steps to you, since I don't know the details on how your CI is defined.

**Comment by lexaknyazev** [comment]
Sounds great! I'll review and finish integration early next week.

**Comment by lexaknyazev** [comment]
It seems that NuGet lacks a straightforward way to provide multiple native binaries with a single .NET wrapper, so a few changes were made to the package.

Instead of copying validator's executable into the output folder of every project that depends on the package, now the wrapper library unpacks proper executable from embedded resources on demand. This should make it possible to use the same DLL on all platforms. 

Among other changes:
- Class used for deserialization does not contain any functions anymore.
- More exceptions are caught in advance with more sound messages.
- Async API is provided so that a single `ValidationContext` instance can handle multiple validation requests at the same time. See examples in the README for more info.

File: nuget/ValidationContext.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+            _embeddedValidatorPath = Path.Combine(Path.GetTempPath(), $"{Path.GetRandomFileName()}_{exeName}");
+
+            using (var fileStream = File.OpenWrite(_embeddedValidatorPath))
+            {
+                resourceStream.CopyTo(fileStream);
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                Process.Start("chmod", $"+x {_embeddedValidatorPath}");
```
Comment (bghgary) [inline nuget/ValidationContext.cs]:
```suggestion
                Process.Start("chmod", $"+x \"{_embeddedValidatorPath}\"");
```

File: nuget/ValidationContext.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        /// <exception cref="InvalidOperationException">
+        ///     Thrown when the <c>ValidationContext</c> was created without embedded
+        ///     validator but custom one was not provided.
+        /// </exception>
+        /// <exception cref="TimeoutException">Thrown when validation takes longer than allowed time.</exception>
+        /// <exception cref="ApplicationException">Thrown when the validator encounters unhandled runtime error.</exception>
+        public ValidationReport Validate(string path, string customValidatorPath = "", bool validateResources = true,
+            TimeSpan timeout = default)
+        {
+            return ValidateAsync(path, customValidatorPath, validateResources, timeout).Result;
```
Comment (bghgary) [inline nuget/ValidationContext.cs]: Does this work? I think you have to wait for the task to complete.

File: nuget/ValidationContext.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        /// <returns>A <c>ValueTask</c> completing with <c>ValidationReport</c> instance.</returns>
+        /// <exception cref="ArgumentException">Thrown when either asset or custom glTF-Validator path is incorrect.</exception>
+        /// <exception cref="FileNotFoundException">Thrown when glTF asset or custom glTF-Validator is not found.</exception>
+        /// <exception cref="InvalidOperationException">
+        ///     Thrown when the <c>ValidationContext</c> was created without embedded
+        ///     validator but custom one was not provided.
+        /// </exception>
+        /// <exception cref="TimeoutException">Thrown when validation takes longer than allowed time.</exception>
+        /// <exception cref="ApplicationException">Thrown when the validator encounters unhandled runtime error.</exception>
+        public async ValueTask<ValidationReport> ValidateAsync(string path, string customValidatorPath = "",
```
Comment (bghgary) [inline nuget/ValidationContext.cs]: I'm not sure why ValueTask is returned here. It probably makes more sense to return Task<ValidationReport> and call Wait on it in the synchronous version.

**Comment by vpenades** [comment]
> I've never seen anyone extract resources into an executable in a temp folder and then run it. I'm not sure what the security implications are.

My original idea was that the command line exe is included as part of your application, not copied to tmp...

The only other alternative would be to compile the cli as a DLL and not as an EXE, and call it with p/invoke, but I have no idea how to do that or if it's even possible for Dart to compile a native dll.

Certainly it's not common, but it's not that uncommon, I've seen a few nugets around that have no other choice that include an exe that does the stuff.

I guess it's fine with due warning. Personally, I want to use it in several unit tests I have around.
My point is that any help you give to developers to validate the exported models will help improve the whole ecosystem as a whole.

File: nuget/ValidationContext.cs (hunk 2 above)
Comment (lexaknyazev) [reply nuget/ValidationContext.cs]: AFAIU, calling `.Result` on `ValueTask` should handle that internally. Maybe need to double-check.

**Comment by lexaknyazev** [comment]
>  command line exe is included as part of your application

I am quite concerned about force-adding a new `.exe` to someone's app folder. Nevertheless, it's still possible to build the .NET wrapper without embedded resources (just leave `compiled` folder empty) and always set a path to app-provided validator.

> I've never seen anyone extract resources into an executable in a temp folder and then run it. I'm not sure what the security implications are.

I think this pattern is quite common among software installers or wrappers for "portable" apps. It seems more secure than running executable that is already present on the disk and could have been replaced.

**Comment by vpenades** [comment]
> I am quite concerned about force-adding a new `.exe` to someone's app folder. Nevertheless, it's still possible to build the .NET wrapper without embedded resources (just leave `compiled` folder empty) and always set a path to app-provided validator.

Just for the sake of reference, I did a small experiment; I went to my nuget local cache directory (typically located in `Users\[You]\.nuget\packages` and I've run a search of all the *.exe files...  and there's quite a few embedded executable CLI tools inside packages

To name some packages that embed executables:
- `microsoft.codecoverage`
- `microsoft.servicefabric.actors`
- `microsoft.testplatform.testhost`
- `xunit.runner.console`
- `geckofx60.32`
- `eo.webbrowser`
- `wix`
- `mono.apitools`
- `mono.webassembly.build`
- `pdfbox`
- 'diskutils`
- 'fbxsharp`

Notice that all these are "normal" packages, not dotnet tool packages.

So I don't think it's that uncommon to embed tools inside packages; and I understand the security concerns, but it's the gltf-validator tool you're including, not some obscure third party exe.

So my opinion is that, for the sake of ease of use, I would preffer to keep the exe files within the package. This will also be specially helpful when automatically updating the nuget package, which will automatically bring new versions of the actual gltf-validator.... otherwise, you need to update the nuget package, and manually update the exe files.

File: nuget/ValidationContext.cs (hunk 2 above)
Comment (bghgary) [reply nuget/ValidationContext.cs]: In the Remarks of the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=netcore-3.1#remarks), it states:

> The following operations should never be performed on a ValueTask<TResult> instance:
> ...
> Using .Result ... when the operation hasn't yet completed...
> ...

Comment (bghgary) [reply nuget/ValidationContext.cs]: Also
> consumers may not read Result until the instance has completed

Comment (lexaknyazev) [reply nuget/ValidationContext.cs]: Oops, it's `Task.Result` that waits until completion, not `ValueTask.Result`.
> Accessing the property's get accessor blocks the calling thread until the asynchronous operation is complete; it is equivalent to calling the `Wait` method.

**Comment by vpenades** [comment]
As an additional note about whether embedding the Exe files inside the nuget packages (which is my preferred choice), the final decision is up to you, but if you choose to not embed the exe in the package, requiring the developers to manually update the exe files along with the nuget package....

I would propose that the nuget code to check the version of the gltf-validator tool to be in line with the one of the nuget package.  My fear is that people unaware of this detail may update the nuget package and keep an old version of the gltf-validator around.

**Comment by vpenades** [comment]
What's the status of this PR? I guess everybody's in containment... ?

**Comment by lexaknyazev** [comment]
Sorry, the PR got auto-closed. If it's still relevant please reopen.

**Comment by vpenades** [comment]
Well... it all depends on whether you're still interested in officially supporting a release of glTF validator as a nuget.

Right now I'm using glTF validator for unit testing directly (using essentially the same calls and logic that would be added to the nuget project). Now that you've published a new release, I'll manually update my project... but it could have been much easier if I would had so simply update a nuget package. This is not a big deal given the current rate of releases (2, 3 per year), but if rate increases is a hassle.

The other reasons for having a nuget package is for other users using c# to easily add additonal validations to their projects, where these scenarios might apply:
- importing models from unknown sources, and the importer's validation is not as robust as glTF validation.
- Testing while developing importers/exporters.
- Enabling strict mode on exporters. (this could be useful for max and blender exporters, to systematically call glTF validator on export)

So again, it all boils down to whether you might want to follow that path or not...

**Pull request closed without merging**